### PR TITLE
JWT authentication

### PR DESCRIPTION
**Pull request by bdunogier**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31875](https://jira.ez.no/browse/EZP-31875)
| **Requires* | ezsystems/ezplatform-admin-ui#1488
| **Type**| feature
| **Target version** | GraphQL `2.2.0` (eZ Platform `v3.2.0`)
| **BC breaks**      | no
| **Tests pass**     | yes/no
| **Doc needed**     | yes

Requirements:

config/packages/security.yaml:
```
security:
    firewalls:
        ezpatform_graphql:
            path: "^/graphql"
            user_checker: eZ\Publish\Core\MVC\Symfony\Security\UserChecker
            anonymous: ~
            guard:
                authenticators:
                    - lexik_jwt_authentication.jwt_token_authenticator
                entry_point: lexik_jwt_authentication.jwt_token_authenticator
            stateless: true
```

config/packages/lexik_jwt_authentication:
```
lexik_jwt_authentication:
    token_extractors:
        authorization_header:
            enabled: true
```

### TODO
- [ ] Open PR on meta-repository with new firewall rule

### CODE REVIEW

File: src/GraphQL/Mutation/Authentication.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    public function __construct(UserService $userService, JWTTokenManagerInterface $tokenManager)
+    {
+        $this->userService = $userService;
+        $this->tokenManager = $tokenManager;
+    }
+
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     */
+    public function createToken($args)
```
Comment (ViniTou) [inline src/GraphQL/Mutation/Authentication.php:36]:
```suggestion
    public function createToken($args): array
```

File: src/GraphQL/Mutation/Authentication.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        $username = $args['username'];
+        $password = $args['password'];
+
+        $apiUser = $this->userService->loadUserByLogin($username);
+        if (!$this->userService->checkUserCredentials($apiUser, $password)) {
+            return ['message' => 'Wrong username or password', 'token' => null];
```
Comment (ViniTou) [inline src/GraphQL/Mutation/Authentication.php:43]: Is that message should be visible to end user on frontend? If yes, we also need translator interface here.

Comment (bdunogier) [reply src/GraphQL/Mutation/Authentication.php:43]: Not necessarily, the frontend can do whatever it wants with it.

File: src/GraphQL/Mutation/Authentication.php (hunk 1 above)
Comment (bdunogier) [reply src/GraphQL/Mutation/Authentication.php:36]: Good point.

File: src/GraphQL/Mutation/Authentication.php (hunk 2 above)
Comment (ViniTou) [reply src/GraphQL/Mutation/Authentication.php:43]: Ok, I jsut remember that in those cases it is really cumbersome if additional case appears - without, for example some status code, frontend would have to parse message from this and then fetch new one for frontend

**Review by ViniTou — APPROVED** [review]

**Review by kmadejski — APPROVED** [review]

**Comment by bdunogier** [comment]
@ezsystems/documentation-team 

### How to authenticate
```
mutation login {
  createToken(username: "admin", password: "publish") {
    token
    message
  }
}
```

Will return a json token if successful:
```
{
  "data": {
    "createToken": {
      "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpYXQiOjE2MDE5NzQ0NjQsImV4cCI6MTYwMTk3ODA2NCwicm9sZXMiOlsiUk9MRV9VU0VSIl0sInVzZXJuYW1lIjoiYWRtaW4ifQ.njTPaZcit0E3yJsGc3k1GXbAfC7RrKgH-z_rPt1z2yg",
      "message": null
    }
  }
}
```

If authentication fails, no token is returned, and a message specifies why.

### Important
JWT authentication is NOT available for the admin siteaccess.

**Review by adamwojs — APPROVED** [review]

File: src/GraphQL/Mutation/Authentication.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public function __construct(UserService $userService, JWTTokenManagerInterface $tokenManager)
+    {
+        $this->userService = $userService;
+        $this->tokenManager = $tokenManager;
+    }
+
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     */
+    public function createToken($args): array
```
[the remark this answers is not in the document]
Comment (bdunogier) [reply src/GraphQL/Mutation/Authentication.php:36]: No, it isn't an array, it is an Argument object that implements ArrayAccess.

**Comment by bdunogier** [comment]
yay

**Comment by micszo** [comment]
There's a difference in the response between invalid password and invalid username.

Invalid password response looks like this:
`
{
  "data": {
    "createToken": {
      "token": null,
      "message": "Wrong username or password"
    }
  }
}
`

But invalid username leads to a 500:
`
{
  "errors": [
    {
      "debugMessage": "Could not find 'user' with identifier 'adminn'",
      "message": "Internal server Error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "createToken"
      ],
      "trace": [
        {
          "file": "/Users/michalszoltysek/Projects/workspace/ez5.lh/var/cache/dev/ContainerRjHIcHD/Handler_65f7bf7.php",
          "line": 44,
          "call": "eZ\\Publish\\Core\\Persistence\\Legacy\\User\\Handler::loadByLogin('adminn')"
        },
(...)
`

**Pull request merged**